Mypy reports many errors, and they should be fixed

### Describe the issue

Unsetting the "ignore_errors" flag for mypy in the config file reveals there are many mypy type-checking errors. These should be fixed properly.

<details>

[31msrc/openfermion/ops/operators/symbolic_operator.py:123: error: Incompatible types in assignment (expression has type "None", base class "object" defined the type as "Callable[[], int]")  [assignment]
src/openfermion/resource_estimates/surface_code_compilation/physical_costing.py:86: error: Argument "failure_rate" to "MagicStateFactory" has incompatible type "float"; expected "int"  [arg-type]
src/openfermion/resource_estimates/surface_code_compilation/physical_costing.py:128: error: Argument "rounds" to "MagicStateFactory" has incompatible type "float"; expected "int"  [arg-type]
src/openfermion/resource_estimates/surface_code_compilation/physical_costing.py:129: error: Argument "failure_rate" to "MagicStateFactory" has incompatible type "float"; expected "int"  [arg-type]
src/openfermion/resource_estimates/pbc/thc/factorizations/kmeans.py:139: error: Incompatible types in assignment (expression has type "floating[Any]", variable has type "float")  [assignment]
src/openfermion/resource_estimates/pbc/hamiltonian/hamiltonian.py:115: error: Incompatible types in assignment (expression has type "type[complex128]", variable has type "type[floating[_64Bit]]")  [assignment]
src/openfermion/resource_estimates/pbc/hamiltonian/hamiltonian.py:116: error: Incompatible types in assignment (expression has type "dtype[Any]", variable has type "type[floating[_64Bit]]")  [assignment]
src/openfermion/resource_estimates/pbc/hamiltonian/hamiltonian.py:127: error: Need type annotation for "tao" (hint: "tao: list[<type>] = ...")  [var-annotated]
src/openfermion/ops/representations/polynomial_tensor.py:120: error: Incompatible types in assignment (expression has type "None", base class "object" defined the type as "Callable[[], int]")  [assignment]
src/openfermion/linalg/wave_fitting.py:71: error: Incompatible types in assignment (expression has type "tuple[Any, ...]", variable has type "ndarray[Any, Any]")  [assignment]
src/openfermion/contrib/representability/_namedtensor.py:46: error: Incompatible types in assignment (expression has type "None", variable has type "int")  [assignment]
src/openfermion/contrib/representability/_namedtensor.py:47: error: Incompatible types in assignment (expression has type "None", variable has type "ndarray[tuple[int, ...], dtype[Any]]")  [assignment]
src/openfermion/contrib/representability/_namedtensor.py:49: error: Incompatible types in assignment (expression has type "Bijection | None", variable has type "Bijection")  [assignment]
src/openfermion/contrib/representability/_namedtensor.py:185: error: Argument "order" to "reshape" has incompatible type "str | None"; expected "Literal['A', 'C', 'F'] | None"  [arg-type]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:24: error: Argument "tensor_elements" to "DualBasisElement" has incompatible type "list[tuple[int, int, int, int]]"; expected "list[tuple[int, ...]] | None"  [arg-type]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:24: note: "list" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:24: note: Consider using "Sequence" instead, which is covariant
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:27: error: Argument "scalar" to "DualBasisElement" has incompatible type "float"; expected "int | None"  [arg-type]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:107: error: Incompatible types in assignment (expression has type "float", variable has type "int | None")  [assignment]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:111: error: Incompatible types in assignment (expression has type "float", variable has type "int | None")  [assignment]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:134: error: Incompatible types in assignment (expression has type "float | int", variable has type "int | None")  [assignment]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:154: error: Incompatible types in assignment (expression has type "float | int", variable has type "int | None")  [assignment]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:254: error: Unsupported operand types for * ("float" and "None")  [operator]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:254: note: Right operand is of type "float | int | None"
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:255: error: Unsupported operand types for * ("float" and "None")  [operator]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:255: note: Right operand is of type "float | int | None"
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:256: error: Unsupported operand types for * ("float" and "None")  [operator]
src/openfermion/contrib/representability/constraints/spin_orbital_2pos_constraints.py:256: note: Right operand is of type "float | int | None"
src/openfermion/transforms/opconversions/binary_codes.py:137: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
src/openfermion/transforms/opconversions/binary_codes.py:138: error: Argument 1 to "len" has incompatible type "int"; expected "Sized"  [arg-type]
src/openfermion/transforms/opconversions/binary_codes.py:140: error: Value of type "int" is not indexable  [index]
src/openfermion/transforms/opconversions/binary_codes.py:142: error: No overload variant of "list" matches argument type "int"  [call-overload]
src/openfermion/transforms/opconversions/binary_codes.py:142: note: Possible overload variants:
src/openfermion/transforms/opconversions/binary_codes.py:142: note:     def [_T] __init__(self) -> list[_T]
src/openfermion/transforms/opconversions/binary_codes.py:142: note:     def [_T] __init__(self, Iterable[_T], /) -> list[_T]
src/openfermion/transforms/opconversions/binary_codes.py:229: error: No overload variant of "__setitem__" of "list" matches argument types "Any", "BinaryPolynomial"  [call-overload]
src/openfermion/transforms/opconversions/binary_codes.py:229: note: Possible overload variants:
src/openfermion/transforms/opconversions/binary_codes.py:229: note:     def __setitem__(self, SupportsIndex, int, /) -> None
src/openfermion/transforms/opconversions/binary_codes.py:229: note:     def __setitem__(self, slice[Any, Any, Any], Iterable[int], /) -> None
src/openfermion/transforms/opconversions/binary_code_transform.py:38: error: Incompatible types in assignment (expression has type "QubitOperator", variable has type "int")  [assignment]
src/openfermion/transforms/opconversions/binary_code_transform.py:42: error: Incompatible types in assignment (expression has type "QubitOperator", variable has type "int")  [assignment]
src/openfermion/transforms/opconversions/binary_code_transform.py:45: error: Incompatible return value type (got "int", expected "QubitOperator")  [return-value]
src/openfermion/hamiltonians/special_operators.py:247: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
src/openfermion/hamiltonians/special_operators.py:249: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
src/openfermion/hamiltonians/special_operators.py:252: error: Incompatible types in assignment (expression has type "tuple[int, str]", variable has type "tuple[int, int] | str | None")  [assignment]
src/openfermion/hamiltonians/special_operators.py:262: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
src/openfermion/hamiltonians/special_operators.py:301: error: Incompatible types in assignment (expression has type "type[BosonOperator]", variable has type "type[FermionOperator]")  [assignment]
src/openfermion/hamiltonians/jellium.py:79: error: List item 0 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:79: error: List item 1 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:124: error: List item 0 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:124: error: List item 1 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:130: error: Need type annotation for "shifted_indices_minus_dict" (hint: "shifted_indices_minus_dict: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:131: error: Need type annotation for "shifted_indices_plus_dict" (hint: "shifted_indices_plus_dict: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:132: error: Need type annotation for "orbital_ids" (hint: "orbital_ids: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:233: error: List item 0 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:233: error: List item 1 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/jellium.py:241: error: Need type annotation for "orbital_ids" (hint: "orbital_ids: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:274: error: Need type annotation for "orbital_a" (hint: "orbital_a: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:275: error: Need type annotation for "orbital_b" (hint: "orbital_b: dict[<type>, <type>] = ...")  [var-annotated]
src/openfermion/hamiltonians/jellium.py:301: error: Incompatible types in assignment (expression has type "tuple[tuple[Any, int], tuple[Any, int], tuple[Any, int], tuple[Any, int]]", variable has type "tuple[tuple[Any, int], tuple[Any, int]]")  [assignment]
src/openfermion/resource_estimates/thc/utils/thc_objectives.py:22: error: Unsupported operand types for - ("None" and "int")  [operator]
src/openfermion/resource_estimates/thc/utils/thc_objectives.py:22: note: Left operand is of type "int | None"
src/openfermion/resource_estimates/thc/utils/thc_factorization.py:28: error: Unsupported operand types for - ("None" and "int")  [operator]
src/openfermion/resource_estimates/thc/utils/thc_factorization.py:28: note: Left operand is of type "int | None"
src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py:66: error: Incompatible return value type (got "floating[Any]", expected "float")  [return-value]
src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py:256: error: Incompatible return value type (got "tuple[dict[tuple[int, int, int], int], ndarray[tuple[int, int], dtype[float64]]]", expected "ndarray[tuple[int, ...], dtype[Any]]")  [return-value]
src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py:383: error: Incompatible return value type (got "tuple[ndarray[tuple[int, int], dtype[signedinteger[_32Bit]]], ndarray[tuple[int, int], dtype[signedinteger[_32Bit]]]]", expected "ndarray[tuple[int, ...], dtype[Any]]")  [return-value]
src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py:585: error: Incompatible types in assignment (expression has type "list[ndarray[tuple[int, ...], dtype[Any]]]", variable has type "ndarray[tuple[int, ...], dtype[Any]]")  [assignment]
src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py:954: error: Argument 2 to "interp_indx_from_qrcp" has incompatible type "int"; expected "ndarray[tuple[int, ...], dtype[Any]]"  [arg-type]
src/openfermion/resource_estimates/pbc/resources/data_types.py:73: error: Argument 1 to "json_normalize" has incompatible type "Series[Any]"; expected "dict[str, Any] | list[dict[str, Any]]"  [arg-type]
src/openfermion/resource_estimates/pbc/resources/data_types.py:74: error: Argument 1 to "json_normalize" has incompatible type "Series[Any]"; expected "dict[str, Any] | list[dict[str, Any]]"  [arg-type]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:150: error: Incompatible return value type (got "tuple[ndarray[tuple[int, ...], dtype[complex128]], list[ndarray[tuple[int, ...], dtype[complex128]]]]", expected "tuple[ndarray[tuple[int, ...], dtype[Any]], ndarray[tuple[int, ...], dtype[Any]]]")  [return-value]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:199: error: Item "int" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:199: error: Item "float" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:199: error: Item "complex" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:202: error: Item "int" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:202: error: Item "float" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:202: error: Item "complex" of "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | int | float | complex" has no attribute "conj"  [union-attr]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:217: error: Incompatible return value type (got "Array", expected "float")  [return-value]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:266: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:266: error: Invalid index type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" for "ndarray[Any, Any]"; expected type "ndarray[tuple[int, ...], dtype[integer[Any] | numpy.bool[builtins.bool]]] | tuple[ndarray[tuple[int, ...], dtype[integer[Any] | numpy.bool[builtins.bool]]], ...]"  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:283: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:283: error: Invalid index type "tuple[Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex, Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex]" for "ndarray[Any, Any]"; expected type "ndarray[tuple[int, ...], dtype[integer[Any] | numpy.bool[builtins.bool]]] | tuple[ndarray[tuple[int, ...], dtype[integer[Any] | numpy.bool[builtins.bool]]], ...]"  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:320: error: Argument 1 to "unpack_thc_factors" has incompatible type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex"; expected "ndarray[tuple[int, ...], dtype[Any]]"  [arg-type]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:332: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:333: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:334: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:335: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:336: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:337: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:338: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:339: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:341: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:344: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:347: error: Value of type "Array | ndarray[Any, Any] | numpy.bool[builtins.bool] | number[Any, int | float | complex] | builtins.bool | int | float | complex" is not indexable  [index]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:913: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:946: error: Incompatible types in assignment (expression has type "tuple[()]", variable has type "ndarray[tuple[int, ...], dtype[Any]]")  [assignment]
src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py:951: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
src/openfermion/resource_estimates/pbc/sparse/compute_lambda_sparse.py:92: error: Argument "num_sym_unique" to "SparseHamiltonianProperties" has incompatible type "int | tuple[int, int]"; expected "int"  [arg-type]
src/openfermion/resource_estimates/pbc/thc/compute_thc_resources.py:57: error: Argument "beta" to "_compute_cost" has incompatible type "int | None"; expected "int"  [arg-type]
src/openfermion/resource_estimates/pbc/thc/compute_thc_resources.py:196: error: Argument 1 to "QI" has incompatible type "float"; expected "int"  [arg-type]
src/openfermion/transforms/opconversions/bksf.py:343: error: Need type annotation for "operator"  [var-annotated]
src/openfermion/transforms/opconversions/bksf.py:362: error: Need type annotation for "operator"  [var-annotated]
src/openfermion/transforms/opconversions/bksf.py:404: error: Need type annotation for "g"  [var-annotated]
src/openfermion/transforms/opconversions/remove_symmetry_qubits.py:20: error: Name "bravyi_kitaev_tree" already defined (by an import)  [no-redef]
src/openfermion/circuits/vpe_circuits.py:46: error: Argument 1 to "append" of "Circuit" has incompatible type "Gate"; expected "OP_TREE"  [arg-type]
src/openfermion/circuits/vpe_circuits.py:50: error: Argument 1 to "append" of "Circuit" has incompatible type "Gate"; expected "OP_TREE"  [arg-type]
src/openfermion/circuits/vpe_circuits.py:104: error: Item "object" of "object | Any" has no attribute "on"  [union-attr]
src/openfermion/circuits/gates/three_qubit_gates.py:20: error: Incompatible return value type (got "EigenGate", expected "CCZPowGate")  [return-value]
src/openfermion/circuits/gates/common_gates.py:114: error: Incompatible return value type (got "EigenGate", expected "CZPowGate")  [return-value]
src/openfermion/hamiltonians/richardson_gaudin.py:66: error: overloaded function has no attribute "setter"  [attr-defined]
src/openfermion/hamiltonians/richardson_gaudin.py:74: error: overloaded function has no attribute "setter"  [attr-defined]
src/openfermion/hamiltonians/mean_field_dwave.py:85: error: Unsupported operand type for unary - ("float | None")  [operator]
src/openfermion/hamiltonians/mean_field_dwave.py:88: error: Unsupported operand type for unary - ("float | None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:117: error: Argument "constant" to "generate_hamiltonian" has incompatible type "float | None"; expected "float"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:172: error: Argument 1 to "len" has incompatible type "list[int] | None"; expected "Sized"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:178: error: Unsupported operand types for * ("int" and "None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:178: error: Unsupported left operand type for * ("None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:178: note: Both left and right operands are unions
src/openfermion/hamiltonians/hartree_fock.py:182: error: Unsupported operand types for * ("int" and "None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:182: error: Unsupported left operand type for * ("None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:182: note: Both left and right operands are unions
src/openfermion/hamiltonians/hartree_fock.py:186: error: Argument 1 to "len" has incompatible type "list[int] | None"; expected "Sized"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:200: error: Item "None" of "InteractionOperator | None" has no attribute "one_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:211: error: Item "None" of "InteractionOperator | None" has no attribute "one_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:222: error: Item "None" of "InteractionOperator | None" has no attribute "two_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:233: error: Item "None" of "InteractionOperator | None" has no attribute "two_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:244: error: Item "None" of "InteractionOperator | None" has no attribute "two_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:255: error: Item "None" of "InteractionOperator | None" has no attribute "two_body_tensor"  [union-attr]
src/openfermion/hamiltonians/hartree_fock.py:350: error: Incompatible types in assignment (expression has type "range", variable has type "list[int] | None")  [assignment]
src/openfermion/hamiltonians/hartree_fock.py:352: error: Incompatible types in assignment (expression has type "range", variable has type "list[int] | None")  [assignment]
src/openfermion/hamiltonians/hartree_fock.py:358: error: Argument 1 to "len" has incompatible type "list[int] | None"; expected "Sized"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:359: error: Argument 1 to "product" has incompatible type "list[int] | None"; expected "Iterable[int]"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:359: error: Argument 2 to "product" has incompatible type "list[int] | None"; expected "Iterable[int]"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:382: error: Need type annotation for "initial_opdm"  [var-annotated]
src/openfermion/hamiltonians/hartree_fock.py:382: error: Argument 1 to "diag" has incompatible type "list[int]"; expected "_SupportsArray[dtype[Never]] | _NestedSequence[_SupportsArray[dtype[Never]]]"  [arg-type]
src/openfermion/hamiltonians/hartree_fock.py:382: error: Unsupported operand types for * ("list[int]" and "int | None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:382: error: Unsupported operand types for * ("list[int]" and "Any | None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:432: error: Too many values to unpack (3 expected, 4 provided)  [misc]
src/openfermion/hamiltonians/hartree_fock.py:434: error: Unsupported operand types for * ("int" and "None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:434: error: Unsupported left operand type for * ("None")  [operator]
src/openfermion/hamiltonians/hartree_fock.py:434: note: Both left and right operands are unions
src/openfermion/hamiltonians/plane_wave_hamiltonian.py:62: error: List item 0 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/plane_wave_hamiltonian.py:62: error: List item 1 has incompatible type "int"; expected "None"  [list-item]
src/openfermion/hamiltonians/plane_wave_hamiltonian.py:88: error: Incompatible return value type (got "FermionOperator | None", expected "FermionOperator")  [return-value]
src/openfermion/hamiltonians/__init__.py:57: error: Incompatible import of "number_operator" (imported name has type "Callable[[int, int | None, Any, int], BosonOperator | FermionOperator]", local name has type "Callable[[Any, Any, Any], Any]")  [assignment]
src/openfermion/circuits/gates/fermionic_simulation.py:134: error: Incompatible types in assignment (expression has type "tuple[int, int, int]", variable has type "tuple[int, int]")  [assignment]
src/openfermion/circuits/gates/fermionic_simulation.py:140: error: Incompatible types in assignment (expression has type "tuple[int, int, int, int]", variable has type "tuple[int, int]")  [assignment]
src/openfermion/circuits/primitives/state_preparation.py:30: error: Missing return statement  [return]
src/openfermion/circuits/primitives/state_preparation.py:84: error: Missing return statement  [return]
src/openfermion/circuits/primitives/state_preparation.py:110: error: Missing return statement  [return]
src/openfermion/circuits/primitives/state_preparation.py:145: error: Missing return statement  [return]
src/openfermion/circuits/primitives/state_preparation.py:212: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:57: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:63: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:64: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:64: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:64: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:65: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:65: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:65: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:66: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:66: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:66: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:68: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:73: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:73: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:73: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:80: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:81: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:81: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:81: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:82: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:82: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:82: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:83: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:83: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:83: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:85: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:91: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:100: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:101: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:101: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:101: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:104: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:104: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:104: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:107: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:107: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:107: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:111: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:116: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:116: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:116: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:123: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:124: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:124: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:124: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:127: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:127: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:127: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:130: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:130: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:130: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:134: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:143: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:149: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:150: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:150: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:150: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:151: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:151: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:151: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:152: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:152: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:152: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:154: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:159: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:159: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:159: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:169: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:178: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:182: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:191: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:192: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:192: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:192: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:195: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:195: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:195: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:198: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:198: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:198: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:202: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:207: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:207: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:207: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "one_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:220: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/linear_swap_network.py:229: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:64: error: Name "openfermion" is not defined  [name-defined]
src/openfermion/circuits/trotter/algorithms/split_operator.py:75: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:79: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:81: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:93: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:96: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:97: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:97: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:97: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:99: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:104: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:118: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:126: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:129: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:133: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:137: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:139: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:154: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:157: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:158: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:158: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:158: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:162: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:167: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:184: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:192: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:195: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:199: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:205: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:206: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:206: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:206: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:208: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:213: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:221: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:229: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:238: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:242: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:251: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:252: error: Item "FermionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:252: error: Item "QubitOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:252: error: Item "InteractionOperator" of "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian" has no attribute "two_body"  [union-attr]
src/openfermion/circuits/trotter/algorithms/split_operator.py:256: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:261: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:270: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/split_operator.py:281: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/split_operator.py:290: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:92: error: Argument 1 to "AsymmetricLowRankTrotterStep" has incompatible type "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian"; expected "InteractionOperator"  [arg-type]
src/openfermion/circuits/trotter/algorithms/low_rank.py:97: error: Argument 1 to "ControlledAsymmetricLowRankTrotterStep" has incompatible type "FermionOperator | QubitOperator | InteractionOperator | DiagonalCoulombHamiltonian"; expected "InteractionOperator"  [arg-type]
src/openfermion/circuits/trotter/algorithms/low_rank.py:148: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/low_rank.py:154: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:171: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:174: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:176: error: Argument "rads" to "rot11" has incompatible type "ndarray[tuple[int, ...], dtype[float64]]"; expected "float"  [arg-type]
src/openfermion/circuits/trotter/algorithms/low_rank.py:188: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:200: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/low_rank.py:210: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:214: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/low_rank.py:223: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:240: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:243: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:245: error: Argument 1 to "rot111" has incompatible type "ndarray[tuple[int, ...], dtype[float64]]"; expected "float"  [arg-type]
src/openfermion/circuits/trotter/algorithms/low_rank.py:261: error: Module not callable  [operator]
src/openfermion/circuits/trotter/algorithms/low_rank.py:276: error: Missing return statement  [return]
src/openfermion/circuits/trotter/algorithms/low_rank.py:286: error: Module not callable  [operator]
src/openfermion/circuits/primitives/bogoliubov_transform.py:23: error: Missing return statement  [return]
src/openfermion/circuits/primitives/bogoliubov_transform.py:115: error: Return value expected  [return-value]
src/openfermion/circuits/primitives/bogoliubov_transform.py:141: error: Missing return statement  [return]
src/openfermion/circuits/primitives/bogoliubov_transform.py:170: error: Missing return statement  [return]
src/openfermion/circuits/primitives/bogoliubov_transform.py:203: error: Missing return statement  [return]
src/openfermion/circuits/primitives/ffft.py:242: error: Argument 2 to "_permute" has incompatible type "list[signedinteger[_64Bit]]"; expected "list[int]"  [arg-type]
src/openfermion/circuits/primitives/ffft.py:269: error: No overload variant of "__getitem__" of "Sequence" matches argument type "signedinteger[_64Bit]"  [call-overload]
src/openfermion/circuits/primitives/ffft.py:269: note: Possible overload variants:
src/openfermion/circuits/primitives/ffft.py:269: note:     def __getitem__(self, int, /) -> Qid
src/openfermion/circuits/primitives/ffft.py:269: note:     def __getitem__(self, slice[Any, Any, Any], /) -> Sequence[Qid]
src/openfermion/circuits/trotter/simulate_trotter.py:27: error: Missing return statement  [return]
src/openfermion/circuits/trotter/simulate_trotter.py:114: error: Missing return statement  [return]
src/openfermion/measurements/fermion_partitioning.py:19: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
src/openfermion/measurements/fermion_partitioning.py:84: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
src/openfermion/measurements/fermion_partitioning.py:200: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
src/openfermion/measurements/fermion_partitioning.py:233: error: Need type annotation for "pairing"  [var-annotated]
src/openfermion/measurements/fermion_partitioning.py:370: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
src/openfermion/measurements/fermion_partitioning.py:413: error: List comprehension has incompatible type List[list[Any]]; expected List[tuple[Any, ...]]  [misc]
src/openfermion/measurements/fermion_partitioning.py:429: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
Found 304 errors in 38 files (checked 352 source files)
[0m
</details>


### What version of this software are you using?

1.7.x

### How can the issue be reproduced?

1. Remove the setting from `dev_tools/conf/mypy.ini`
2. Run `check/mypy`

--- a/dev_tools/shell_tools_test.py
+++ b/dev_tools/shell_tools_test.py
@@ -28,13 +28,11 @@ def only_on_posix(func: Callable) -> Optional[Callable]:
 
 
 def run_cmd(*args: Any, **kwargs: Any) -> shell_tools.CommandOutput:
-    kwargs = dict(kwargs)
     kwargs['log_run_to_stderr'] = False
     return shell_tools.run_cmd(*args, **kwargs)
 
 
 def run_shell(*args: Any, **kwargs: Any) -> shell_tools.CommandOutput:
-    kwargs = dict(kwargs)
     kwargs['log_run_to_stderr'] = False
     return shell_tools.run_shell(*args, **kwargs)
 

--- a/dev_tools/shell_tools_test.py
+++ b/dev_tools/shell_tools_test.py
@@ -28,11 +28,15 @@ def only_on_posix(func: Callable) -> Optional[Callable]:
 
 
 def run_cmd(*args: Any, **kwargs: Any) -> shell_tools.CommandOutput:
-    return shell_tools.run_cmd(*args, log_run_to_stderr=False, **kwargs)
+    kwargs = dict(kwargs)
+    kwargs['log_run_to_stderr'] = False
+    return shell_tools.run_cmd(*args, **kwargs)
 
 
 def run_shell(*args: Any, **kwargs: Any) -> shell_tools.CommandOutput:
-    return shell_tools.run_shell(*args, log_run_to_stderr=False, **kwargs)
+    kwargs = dict(kwargs)
+    kwargs['log_run_to_stderr'] = False
+    return shell_tools.run_shell(*args, **kwargs)
 
 
 @only_on_posix

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,8 +33,7 @@ filterwarnings = [
 
 [tool.mypy]
 implicit_optional = true
-# TODO: this is temporary for 1.7.0 release. Remove this & fix the errors.
-ignore_errors = true
+ignore_errors = false
 
 [[tool.mypy.overrides]]
 module = "__main__"

--- a/src/openfermion/circuits/trotter/algorithms/linear_swap_network.py
+++ b/src/openfermion/circuits/trotter/algorithms/linear_swap_network.py
@@ -11,13 +11,13 @@
 #   limitations under the License.
 """A Trotter algorithm using the "fermionic simulation gate"."""
 
-from typing import cast, Optional, Sequence, Tuple
+from typing import cast, Iterator, Optional, Sequence, Tuple
 
 import cirq
 
 from openfermion.circuits.gates import Ryxxy, Rxxyy, CRyxxy, CRxxyy, rot111, rot11
 from openfermion.ops import DiagonalCoulombHamiltonian
-from openfermion.circuits.primitives import swap_network
+from openfermion.circuits.primitives.swap_network import swap_network
 from openfermion.circuits.trotter.trotter_algorithm import (
     Hamiltonian,
     TrotterStep,
@@ -38,73 +38,84 @@ class LinearSwapNetworkTrotterAlgorithm(TrotterAlgorithm):
     supported_types = {DiagonalCoulombHamiltonian}
 
     def symmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return SymmetricLinearSwapNetworkTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, hamiltonian)
+        return SymmetricLinearSwapNetworkTrotterStep(dc_hamiltonian)
 
     def asymmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return AsymmetricLinearSwapNetworkTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, hamiltonian)
+        return AsymmetricLinearSwapNetworkTrotterStep(dc_hamiltonian)
 
     def controlled_symmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return ControlledSymmetricLinearSwapNetworkTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, hamiltonian)
+        return ControlledSymmetricLinearSwapNetworkTrotterStep(dc_hamiltonian)
 
     def controlled_asymmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return ControlledAsymmetricLinearSwapNetworkTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, hamiltonian)
+        return ControlledAsymmetricLinearSwapNetworkTrotterStep(dc_hamiltonian)
 
 
 LINEAR_SWAP_NETWORK = LinearSwapNetworkTrotterAlgorithm()
 
 
-class SymmetricLinearSwapNetworkTrotterStep(TrotterStep):
+class LinearSwapNetworkTrotterStep(TrotterStep):
+    def __init__(self, hamiltonian: DiagonalCoulombHamiltonian) -> None:
+        super().__init__(hamiltonian)
+
+
+class SymmetricLinearSwapNetworkTrotterStep(LinearSwapNetworkTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, self.hamiltonian)
 
         # Apply one- and two-body interactions for half of the full time
-        def one_and_two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield Rxxyy(0.5 * self.hamiltonian.one_body[p, q].real * time).on(a, b)
-            yield Ryxxy(0.5 * self.hamiltonian.one_body[p, q].imag * time).on(a, b)
-            yield rot11(rads=-self.hamiltonian.two_body[p, q] * time).on(a, b)
+        def one_and_two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield Rxxyy(0.5 * dc_hamiltonian.one_body[p, q].real * time).on(a, b)
+            yield Ryxxy(0.5 * dc_hamiltonian.one_body[p, q].imag * time).on(a, b)
+            yield rot11(rads=float(-dc_hamiltonian.two_body[p, q] * time)).on(a, b)
 
         yield swap_network(qubits, one_and_two_body_interaction, fermionic=True)
         qubits = qubits[::-1]
 
         # Apply one-body potential for the full time
         yield (
-            cirq.rz(rads=-self.hamiltonian.one_body[i, i].real * time).on(qubits[i])
+            cirq.rz(rads=-dc_hamiltonian.one_body[i, i].real * time).on(qubits[i])
             for i in range(n_qubits)
         )
 
         # Apply one- and two-body interactions for half of the full time
         # This time, reorder the operations so that the entire Trotter step is
         # symmetric
-        def one_and_two_body_interaction_reverse_order(p, q, a, b) -> cirq.OP_TREE:
-            yield rot11(rads=-self.hamiltonian.two_body[p, q] * time).on(a, b)
-            yield Ryxxy(0.5 * self.hamiltonian.one_body[p, q].imag * time).on(a, b)
-            yield Rxxyy(0.5 * self.hamiltonian.one_body[p, q].real * time).on(a, b)
+        def one_and_two_body_interaction_reverse_order(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot11(rads=float(-dc_hamiltonian.two_body[p, q] * time)).on(a, b)
+            yield Ryxxy(0.5 * dc_hamiltonian.one_body[p, q].imag * time).on(a, b)
+            yield Rxxyy(0.5 * dc_hamiltonian.one_body[p, q].real * time).on(a, b)
 
         yield swap_network(
             qubits, one_and_two_body_interaction_reverse_order, fermionic=True, offset=True
         )
 
 
-class ControlledSymmetricLinearSwapNetworkTrotterStep(TrotterStep):
+class ControlledSymmetricLinearSwapNetworkTrotterStep(LinearSwapNetworkTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, self.hamiltonian)
 
         if not isinstance(control_qubit, cirq.Qid):
             raise TypeError('Control qudit must be specified.')
 
         # Apply one- and two-body interactions for half of the full time
-        def one_and_two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield CRxxyy(0.5 * self.hamiltonian.one_body[p, q].real * time).on(
+        def one_and_two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield CRxxyy(0.5 * dc_hamiltonian.one_body[p, q].real * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield CRyxxy(0.5 * self.hamiltonian.one_body[p, q].imag * time).on(
+            yield CRyxxy(0.5 * dc_hamiltonian.one_body[p, q].imag * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield rot111(-self.hamiltonian.two_body[p, q] * time).on(
+            yield rot111(rads=float(-dc_hamiltonian.two_body[p, q] * time)).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
 
@@ -113,21 +124,21 @@ class ControlledSymmetricLinearSwapNetworkTrotterStep(TrotterStep):
 
         # Apply one-body potential for the full time
         yield (
-            rot11(rads=-self.hamiltonian.one_body[i, i].real * time).on(control_qubit, qubits[i])
+            rot11(rads=-dc_hamiltonian.one_body[i, i].real * time).on(control_qubit, qubits[i])
             for i in range(n_qubits)
         )
 
         # Apply one- and two-body interactions for half of the full time
         # This time, reorder the operations so that the entire Trotter step is
         # symmetric
-        def one_and_two_body_interaction_reverse_order(p, q, a, b) -> cirq.OP_TREE:
-            yield rot111(-self.hamiltonian.two_body[p, q] * time).on(
+        def one_and_two_body_interaction_reverse_order(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot111(rads=float(-dc_hamiltonian.two_body[p, q] * time)).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield CRyxxy(0.5 * self.hamiltonian.one_body[p, q].imag * time).on(
+            yield CRyxxy(0.5 * dc_hamiltonian.one_body[p, q].imag * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield CRxxyy(0.5 * self.hamiltonian.one_body[p, q].real * time).on(
+            yield CRxxyy(0.5 * dc_hamiltonian.one_body[p, q].real * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
 
@@ -136,27 +147,28 @@ class ControlledSymmetricLinearSwapNetworkTrotterStep(TrotterStep):
         )
 
         # Apply phase from constant term
-        yield cirq.rz(rads=-self.hamiltonian.constant * time).on(control_qubit)
+        yield cirq.rz(rads=-dc_hamiltonian.constant * time).on(control_qubit)
 
 
-class AsymmetricLinearSwapNetworkTrotterStep(TrotterStep):
+class AsymmetricLinearSwapNetworkTrotterStep(LinearSwapNetworkTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, self.hamiltonian)
 
         # Apply one- and two-body interactions for the full time
-        def one_and_two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield Rxxyy(self.hamiltonian.one_body[p, q].real * time).on(a, b)
-            yield Ryxxy(self.hamiltonian.one_body[p, q].imag * time).on(a, b)
-            yield rot11(rads=-2 * self.hamiltonian.two_body[p, q] * time).on(a, b)
+        def one_and_two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield Rxxyy(dc_hamiltonian.one_body[p, q].real * time).on(a, b)
+            yield Ryxxy(dc_hamiltonian.one_body[p, q].imag * time).on(a, b)
+            yield rot11(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(a, b)
 
         yield swap_network(qubits, one_and_two_body_interaction, fermionic=True)
         qubits = qubits[::-1]
 
         # Apply one-body potential for the full time
         yield (
-            cirq.rz(rads=-self.hamiltonian.one_body[i, i].real * time).on(qubits[i])
+            cirq.rz(rads=-dc_hamiltonian.one_body[i, i].real * time).on(qubits[i])
             for i in range(n_qubits)
         )
 
@@ -172,30 +184,31 @@ class AsymmetricLinearSwapNetworkTrotterStep(TrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # If the number of Trotter steps is odd, possibly swap qubits back
         if n_steps & 1 and not omit_final_swaps:
             yield swap_network(qubits, fermionic=True)
 
 
-class ControlledAsymmetricLinearSwapNetworkTrotterStep(TrotterStep):
+class ControlledAsymmetricLinearSwapNetworkTrotterStep(LinearSwapNetworkTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(DiagonalCoulombHamiltonian, self.hamiltonian)
 
         if not isinstance(control_qubit, cirq.Qid):
             raise TypeError('Control qudit must be specified.')
 
         # Apply one- and two-body interactions for the full time
-        def one_and_two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield CRxxyy(self.hamiltonian.one_body[p, q].real * time).on(
+        def one_and_two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield CRxxyy(dc_hamiltonian.one_body[p, q].real * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield CRyxxy(self.hamiltonian.one_body[p, q].imag * time).on(
+            yield CRyxxy(dc_hamiltonian.one_body[p, q].imag * time).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
-            yield rot111(-2 * self.hamiltonian.two_body[p, q] * time).on(
+            yield rot111(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
 
@@ -204,12 +217,12 @@ class ControlledAsymmetricLinearSwapNetworkTrotterStep(TrotterStep):
 
         # Apply one-body potential for the full time
         yield (
-            rot11(rads=-self.hamiltonian.one_body[i, i].real * time).on(control_qubit, qubits[i])
+            rot11(rads=-dc_hamiltonian.one_body[i, i].real * time).on(control_qubit, qubits[i])
             for i in range(n_qubits)
         )
 
         # Apply phase from constant term
-        yield cirq.rz(rads=-self.hamiltonian.constant * time).on(control_qubit)
+        yield cirq.rz(rads=-dc_hamiltonian.constant * time).on(control_qubit)
 
     def step_qubit_permutation(
         self, qubits: Sequence[cirq.Qid], control_qubit: Optional[cirq.Qid] = None
@@ -223,7 +236,7 @@ class ControlledAsymmetricLinearSwapNetworkTrotterStep(TrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # If the number of Trotter steps is odd, possibly swap qubits back
         if n_steps & 1 and not omit_final_swaps:
             yield swap_network(qubits, fermionic=True)

--- a/src/openfermion/circuits/trotter/algorithms/split_operator.py
+++ b/src/openfermion/circuits/trotter/algorithms/split_operator.py
@@ -11,7 +11,7 @@
 #   limitations under the License.
 """A Trotter algorithm using a split-operator approach."""
 
-from typing import cast, Optional, Sequence, Tuple
+from typing import cast, Iterator, Optional, Sequence, Tuple
 
 import cirq
 
@@ -20,7 +20,8 @@ from openfermion.circuits.gates import rot11, rot111
 import openfermion.ops as ops
 
 # import openfermion.circuits.primitives as primitives
-from openfermion.circuits.primitives import bogoliubov_transform, swap_network
+from openfermion.circuits.primitives.bogoliubov_transform import bogoliubov_transform
+from openfermion.circuits.primitives.swap_network import swap_network
 from openfermion.circuits.trotter.trotter_algorithm import (
     Hamiltonian,
     TrotterStep,
@@ -45,23 +46,27 @@ class SplitOperatorTrotterAlgorithm(TrotterAlgorithm):
     supported_types = {ops.DiagonalCoulombHamiltonian}
 
     def symmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return SymmetricSplitOperatorTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, hamiltonian)
+        return SymmetricSplitOperatorTrotterStep(dc_hamiltonian)
 
     def asymmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return AsymmetricSplitOperatorTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, hamiltonian)
+        return AsymmetricSplitOperatorTrotterStep(dc_hamiltonian)
 
     def controlled_symmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return ControlledSymmetricSplitOperatorTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, hamiltonian)
+        return ControlledSymmetricSplitOperatorTrotterStep(dc_hamiltonian)
 
     def controlled_asymmetric(self, hamiltonian: Hamiltonian) -> Optional[TrotterStep]:
-        return ControlledAsymmetricSplitOperatorTrotterStep(hamiltonian)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, hamiltonian)
+        return ControlledAsymmetricSplitOperatorTrotterStep(dc_hamiltonian)
 
 
 SPLIT_OPERATOR = SplitOperatorTrotterAlgorithm()
 
 
 class SplitOperatorTrotterStep(TrotterStep):
-    def __init__(self, hamiltonian: 'openfermion.DiagonalCoulombHamiltonian') -> None:
+    def __init__(self, hamiltonian: ops.DiagonalCoulombHamiltonian) -> None:
         quad_ham = ops.QuadraticHamiltonian(hamiltonian.one_body)
         # Get the basis change matrix that diagonalizes the one-body term
         # and associated orbital energies
@@ -74,14 +79,15 @@ class SplitOperatorTrotterStep(TrotterStep):
 class SymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
     def prepare(
         self, qubits: Sequence[cirq.Qid], control_qubits: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # Change to the basis in which the one-body term is diagonal
         yield cirq.inverse(bogoliubov_transform(qubits, self.basis_change_matrix))
 
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, self.hamiltonian)
 
         # Simulate the one-body terms for half of the full time
         yield (
@@ -93,8 +99,8 @@ class SymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         yield bogoliubov_transform(qubits, self.basis_change_matrix)
 
         # Simulate the two-body terms for the full time
-        def two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield rot11(rads=-2 * self.hamiltonian.two_body[p, q] * time).on(a, b)
+        def two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot11(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(a, b)
 
         yield swap_network(qubits, two_body_interaction)
         # The qubit ordering has been reversed
@@ -121,7 +127,7 @@ class SymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # Rotate back to the computational basis
         yield bogoliubov_transform(qubits, self.basis_change_matrix)
         # If the number of Trotter steps is odd, possibly swap qubits back
@@ -132,14 +138,15 @@ class SymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
 class ControlledSymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
     def prepare(
         self, qubits: Sequence[cirq.Qid], control_qubits: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # Change to the basis in which the one-body term is diagonal
         yield cirq.inverse(bogoliubov_transform(qubits, self.basis_change_matrix))
 
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, self.hamiltonian)
 
         if not isinstance(control_qubit, cirq.Qid):
             raise TypeError('Control qudit must be specified.')
@@ -154,8 +161,8 @@ class ControlledSymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         yield bogoliubov_transform(qubits, self.basis_change_matrix)
 
         # Simulate the two-body terms for the full time
-        def two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield rot111(-2 * self.hamiltonian.two_body[p, q] * time).on(
+        def two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot111(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
 
@@ -173,7 +180,7 @@ class ControlledSymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         )
 
         # Apply phase from constant term
-        yield cirq.rz(rads=-self.hamiltonian.constant * time).on(control_qubit)
+        yield cirq.rz(rads=-dc_hamiltonian.constant * time).on(control_qubit)
 
     def step_qubit_permutation(
         self, qubits: Sequence[cirq.Qid], control_qubit: Optional[cirq.Qid] = None
@@ -187,7 +194,7 @@ class ControlledSymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # Rotate back to the computational basis
         yield bogoliubov_transform(qubits, self.basis_change_matrix)
         # If the number of Trotter steps is odd, possibly swap qubits back
@@ -198,12 +205,13 @@ class ControlledSymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
 class AsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, self.hamiltonian)
 
         # Simulate the two-body terms for the full time
-        def two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield rot11(rads=-2 * self.hamiltonian.two_body[p, q] * time).on(a, b)
+        def two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot11(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(a, b)
 
         yield swap_network(qubits, two_body_interaction)
         # The qubit ordering has been reversed
@@ -232,7 +240,7 @@ class AsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # If the number of Trotter steps is odd, possibly swap qubits back
         if n_steps & 1 and not omit_final_swaps:
             yield swap_network(qubits)
@@ -241,15 +249,16 @@ class AsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
 class ControlledAsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
     def trotter_step(
         self, qubits: Sequence[cirq.Qid], time: float, control_qubit: Optional[cirq.Qid] = None
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         n_qubits = len(qubits)
+        dc_hamiltonian = cast(ops.DiagonalCoulombHamiltonian, self.hamiltonian)
 
         if not isinstance(control_qubit, cirq.Qid):
             raise TypeError('Control qudit must be specified.')
 
         # Simulate the two-body terms for the full time
-        def two_body_interaction(p, q, a, b) -> cirq.OP_TREE:
-            yield rot111(-2 * self.hamiltonian.two_body[p, q] * time).on(
+        def two_body_interaction(p, q, a, b) -> Iterator[cirq.OP_TREE]:
+            yield rot111(rads=float(-2 * dc_hamiltonian.two_body[p, q] * time)).on(
                 cast(cirq.Qid, control_qubit), a, b
             )
 
@@ -270,7 +279,7 @@ class ControlledAsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         yield bogoliubov_transform(qubits, self.basis_change_matrix)
 
         # Apply phase from constant term
-        yield cirq.rz(rads=-self.hamiltonian.constant * time).on(control_qubit)
+        yield cirq.rz(rads=-dc_hamiltonian.constant * time).on(control_qubit)
 
     def step_qubit_permutation(
         self, qubits: Sequence[cirq.Qid], control_qubit: Optional[cirq.Qid] = None
@@ -284,7 +293,7 @@ class ControlledAsymmetricSplitOperatorTrotterStep(SplitOperatorTrotterStep):
         n_steps: int,
         control_qubit: Optional[cirq.Qid] = None,
         omit_final_swaps: bool = False,
-    ) -> cirq.OP_TREE:
+    ) -> Iterator[cirq.OP_TREE]:
         # If the number of Trotter steps is odd, possibly swap qubits back
         if n_steps & 1 and not omit_final_swaps:
             yield swap_network(qubits)

--- a/src/openfermion/resource_estimates/pbc/resources/data_types.py
+++ b/src/openfermion/resource_estimates/pbc/resources/data_types.py
@@ -70,8 +70,8 @@ class PBCResources:
     def to_dataframe(self) -> pd.DataFrame:
         """Convert PBCResources instance to pandas DataFrame."""
         df = pd.DataFrame(self.dict())
-        lambdas = pd.json_normalize(df.pop("ham_props"))
-        resources = pd.json_normalize(df.pop("resources"))
+        lambdas = pd.json_normalize(df.pop("ham_props").tolist())
+        resources = pd.json_normalize(df.pop("resources").tolist())
         df = df.join(pd.DataFrame(lambdas))
         df = df.join(pd.DataFrame(resources))
         return df

--- a/src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py
+++ b/src/openfermion/resource_estimates/pbc/thc/factorizations/isdf.py
@@ -62,7 +62,7 @@ def check_isdf_solution(
     lhs = np.einsum("Ri,Rj->Rij", orbitals.conj(), orbitals, optimize=True)
     rhs = np.einsum("mi,mj->mij", interp_orbitals.conj(), interp_orbitals, optimize=True)
     lhs_check = np.einsum("Rm,mij->Rij", xi, rhs, optimize=True)
-    return np.linalg.norm(lhs - lhs_check)
+    return float(np.linalg.norm(lhs - lhs_check))
 
 
 def solve_isdf(orbitals: npt.NDArray, interp_indx: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:

--- a/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
+++ b/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
@@ -520,7 +520,9 @@ def lbfgsb_opt_kpthc_l2reg(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(
+            chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map
+        )
     return np.array(params), loss
 
 
@@ -662,7 +664,9 @@ def lbfgsb_opt_kpthc_l2reg_batched(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(gq)) for gq in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(
+            chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map
+        )
     return np.array(params), loss
 
 
@@ -755,7 +759,9 @@ def adagrad_opt_kpthc_batched(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(x, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(
+            chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map
+        )
     return params, loss
 
 
@@ -933,7 +939,9 @@ def kpoint_thc_via_isdf(
             )
             info["loss_bfgs"] = loss_bfgs
         num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
-        chi_packed, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+        chi_packed, zeta_list = unpack_thc_factors(
+            opt_params, num_thc, num_mo, num_kpts, num_g_per_q
+        )
         chi = np.asarray(chi_packed)
         zeta = np.array(zeta_list, dtype=object)
     if verbose:

--- a/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
+++ b/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
@@ -27,7 +27,6 @@ be used as
 # pylint: disable=wrong-import-position
 import math
 import time
-from typing import Tuple, Union
 
 import h5py
 import numpy as np
@@ -79,7 +78,7 @@ def save_thc_factors(
     chi: npt.NDArray,
     zeta: npt.NDArray,
     gpq_map: npt.NDArray,
-    xi: Union[npt.NDArray, None] = None,
+    xi: npt.NDArray | None = None,
 ) -> None:
     """Write THC factors to file
 
@@ -115,8 +114,8 @@ def get_zeta_size(zeta: npt.NDArray) -> int:
 
 
 def unpack_thc_factors(
-    xcur: npt.NDArray, num_thc: int, num_orb: int, num_kpts: int, num_g_per_q: list
-) -> Tuple[npt.NDArray, npt.NDArray]:
+    xcur: npt.NDArray, num_thc: int, num_orb: int, num_kpts: int, num_g_per_q: list[int]
+) -> tuple[npt.NDArray, list[npt.NDArray]]:
     """Unpack THC factors from flattened array used for reoptimization.
 
     Args:
@@ -172,10 +171,10 @@ def pack_thc_factors(chi: npt.NDArray, zeta: npt.NDArray, buffer: npt.NDArray) -
 
 @jax.jit
 def compute_objective_batched(
-    chis: Tuple[jnpt.ArrayLike, jnpt.ArrayLike, jnpt.ArrayLike, jnpt.ArrayLike],
-    zetas: jnpt.ArrayLike,
-    chols: Tuple[jnpt.ArrayLike, jnpt.ArrayLike],
-    norm_factors: Tuple[jnpt.ArrayLike, jnpt.ArrayLike, jnpt.ArrayLike, jnpt.ArrayLike],
+    chis: tuple[jax.Array, jax.Array, jax.Array, jax.Array],
+    zetas: jax.Array,
+    chols: tuple[jax.Array, jax.Array],
+    norm_factors: tuple[jax.Array, jax.Array, jax.Array, jax.Array],
     num_kpts: int,
     penalty_param: float = 0.0,
 ) -> float:
@@ -214,12 +213,12 @@ def compute_objective_batched(
     lambda_z = jnp.sum(jnp.einsum("jpq->j", 0.5 * jnp.abs(mpq_normalized)) ** 2.0)
 
     res = 0.5 * jnp.sum((jnp.abs(deri)) ** 2) + penalty_param * lambda_z
-    return res
+    return float(res)
 
 
 def prepare_batched_data_indx_arrays(
     momentum_map: npt.NDArray, gpq_map: npt.NDArray
-) -> Tuple[npt.NDArray, npt.NDArray]:
+) -> tuple[npt.NDArray, npt.NDArray]:
     r"""Create arrays to batch over.
 
     Flatten sum_q sum_{k,k_prime} -> sum_q \sum_{indx} and pack momentum
@@ -253,7 +252,7 @@ def prepare_batched_data_indx_arrays(
 
 
 @jax.jit
-def get_batched_data_1indx(array: jnpt.ArrayLike, indx: jnpt.ArrayLike) -> jnpt.ArrayLike:
+def get_batched_data_1indx(array: jax.Array, indx: jax.Array) -> jax.Array:
     """Helper function to extract entries of array given another array.
 
     Args:
@@ -267,9 +266,7 @@ def get_batched_data_1indx(array: jnpt.ArrayLike, indx: jnpt.ArrayLike) -> jnpt.
 
 
 @jax.jit
-def get_batched_data_2indx(
-    array: jnpt.ArrayLike, indxa: jnpt.ArrayLike, indxb: jnpt.ArrayLike
-) -> jnpt.ArrayLike:
+def get_batched_data_2indx(array: jax.Array, indxa: jax.Array, indxb: jax.Array) -> jax.Array:
     """Helper function to extract entries of 2D array given another array
 
     Args:
@@ -284,13 +281,13 @@ def get_batched_data_2indx(
 
 
 def thc_objective_regularized_batched(
-    xcur: jnpt.ArrayLike,
+    xcur: npt.NDArray | jax.Array,
     num_orb: int,
     num_thc: int,
     momentum_map: npt.NDArray,
     gpq_map: npt.NDArray,
-    chol: jnpt.ArrayLike,
-    indx_arrays: Tuple[jnpt.ArrayLike, jnpt.ArrayLike],
+    chol: npt.NDArray | jax.Array,
+    indx_arrays: tuple[npt.NDArray, npt.NDArray],
     batch_size: int,
     penalty_param=0.0,
 ) -> float:
@@ -317,10 +314,13 @@ def thc_objective_regularized_batched(
     """
     num_kpts = momentum_map.shape[0]
     num_g_per_q = [len(np.unique(gq)) for gq in gpq_map]
-    chi, zeta = unpack_thc_factors(xcur, num_thc, num_orb, num_kpts, num_g_per_q)
+    chi_packed, zeta_list = unpack_thc_factors(
+        np.asarray(xcur), num_thc, num_orb, num_kpts, num_g_per_q
+    )
+    chi_jax = jnp.asarray(chi_packed)
     # Normalization factor, no factor of sqrt as there are 4 chis in total when
     # building ERI.
-    norm_kp = jnp.einsum("kpP,kpP->kP", chi.conj(), chi, optimize=True) ** 0.5
+    norm_kp = jnp.einsum("kpP,kpP->kP", chi_jax.conj(), chi_jax, optimize=True) ** 0.5
     num_batches = math.ceil(num_kpts**2 / batch_size)
 
     indx_pqrs, indx_zeta = indx_arrays
@@ -329,22 +329,22 @@ def thc_objective_regularized_batched(
         for ibatch in range(num_batches):
             start = ibatch * batch_size
             end = (ibatch + 1) * batch_size
-            chi_p = get_batched_data_1indx(chi, indx_pqrs[iq, start:end, 0])
-            chi_q = get_batched_data_1indx(chi, indx_pqrs[iq, start:end, 1])
-            chi_r = get_batched_data_1indx(chi, indx_pqrs[iq, start:end, 2])
-            chi_s = get_batched_data_1indx(chi, indx_pqrs[iq, start:end, 3])
+            chi_p = get_batched_data_1indx(chi_jax, indx_pqrs[iq, start:end, 0])
+            chi_q = get_batched_data_1indx(chi_jax, indx_pqrs[iq, start:end, 1])
+            chi_r = get_batched_data_1indx(chi_jax, indx_pqrs[iq, start:end, 2])
+            chi_s = get_batched_data_1indx(chi_jax, indx_pqrs[iq, start:end, 3])
             norm_k1 = get_batched_data_1indx(norm_kp, indx_pqrs[iq, start:end, 0])
             norm_k2 = get_batched_data_1indx(norm_kp, indx_pqrs[iq, start:end, 1])
             norm_k3 = get_batched_data_1indx(norm_kp, indx_pqrs[iq, start:end, 2])
             norm_k4 = get_batched_data_1indx(norm_kp, indx_pqrs[iq, start:end, 3])
             zeta_batch = get_batched_data_2indx(
-                zeta[iq], indx_zeta[iq, start:end, 0], indx_zeta[iq, start:end, 1]
+                jnp.asarray(zeta_list[iq]), indx_zeta[iq, start:end, 0], indx_zeta[iq, start:end, 1]
             )
             chol_batch_pq = get_batched_data_2indx(
-                chol, indx_pqrs[iq, start:end, 0], indx_pqrs[iq, start:end, 1]
+                jnp.asarray(chol), indx_pqrs[iq, start:end, 0], indx_pqrs[iq, start:end, 1]
             )
             chol_batch_rs = get_batched_data_2indx(
-                chol, indx_pqrs[iq, start:end, 2], indx_pqrs[iq, start:end, 3]
+                jnp.asarray(chol), indx_pqrs[iq, start:end, 2], indx_pqrs[iq, start:end, 3]
             )
             objective += compute_objective_batched(
                 (chi_p, chi_q, chi_r, chi_s),
@@ -419,11 +419,11 @@ def lbfgsb_opt_kpthc_l2reg(
     momentum_map: npt.NDArray,
     gpq_map: npt.NDArray,
     chol: npt.NDArray,
-    chkfile_name: Union[str, None] = None,
+    chkfile_name: str | None = None,
     maxiter: int = 150_000,
     disp_freq: int = 98,
-    penalty_param: Union[float, None] = None,
-) -> Tuple[npt.NDArray, float]:
+    penalty_param: float | None = None,
+) -> tuple[npt.NDArray, float]:
     """Least-squares fit of two-electron integral tensors with  L-BFGS-B
 
     Uses l2-regularization of lambda penalty function.
@@ -521,8 +521,8 @@ def lbfgsb_opt_kpthc_l2reg(
     )
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
-        chi, zeta = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi, zeta, gpq_map)
+        chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
+        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
     return np.array(params), loss
 
 
@@ -532,12 +532,12 @@ def lbfgsb_opt_kpthc_l2reg_batched(
     momentum_map: npt.NDArray,
     gpq_map: npt.NDArray,
     chol: npt.NDArray,
-    chkfile_name: Union[str, None] = None,
+    chkfile_name: str | None = None,
     maxiter: int = 150_000,
     disp_freq: int = 98,
-    penalty_param: Union[float, None] = None,
-    batch_size: Union[int, None] = None,
-) -> Tuple[npt.NDArray, float]:
+    penalty_param: float | None = None,
+    batch_size: int | None = None,
+) -> tuple[npt.NDArray, float]:
     """Least-squares fit of two-electron integral tensors with  L-BFGS-B.
 
     Uses l2-regularization of lambda. This version batches over multiple
@@ -663,8 +663,8 @@ def lbfgsb_opt_kpthc_l2reg_batched(
     params = np.array(res.x)
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(gq)) for gq in gpq_map]
-        chi, zeta = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi, zeta, gpq_map)
+        chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
+        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
     return np.array(params), loss
 
 
@@ -680,7 +680,7 @@ def adagrad_opt_kpthc_batched(
     momentum=0.9,
     maxiter=50_000,
     gtol=1.0e-5,
-) -> Tuple[npt.NDArray, float]:
+) -> tuple[npt.NDArray, float]:
     """Adagrad optimization of THC factors.
 
     THC optimization usually starts with BFGS and then is completed with Adagrad
@@ -756,8 +756,8 @@ def adagrad_opt_kpthc_batched(
     )
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
-        chi, zeta = unpack_thc_factors(x, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi, zeta, gpq_map)
+        chi_out, zeta_out = unpack_thc_factors(x, num_thc, num_orb, num_kpts, num_g_per_q)
+        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
     return params, loss
 
 
@@ -836,13 +836,13 @@ def kpoint_thc_via_isdf(
     checkpoint_basename: str = "thc",
     save_checkpoints: bool = False,
     use_batched_algos: bool = True,
-    penalty_param: Union[None, float] = None,
-    batch_size: Union[None, bool] = None,
+    penalty_param: float | None = None,
+    batch_size: int | None = None,
     max_kmeans_iteration: int = 500,
     verbose: bool = False,
-    initial_guess: Union[None, KPointTHC] = None,
+    initial_guess: KPointTHC | None = None,
     isdf_density_guess: bool = False,
-) -> Tuple[KPointTHC, dict]:
+) -> tuple[KPointTHC, dict]:
     """
     Solve k-point THC using ISDF as an initial guess.
 
@@ -897,8 +897,7 @@ def kpoint_thc_via_isdf(
     num_kpts = len(kmf.kpts)
     chi, zeta, g_mapping = kpt_thc.chi, kpt_thc.zeta, kpt_thc.g_mapping
     if save_checkpoints:
-        chkfile_name = f"{checkpoint_basename}_isdf.h5"
-        save_thc_factors(chkfile_name, chi, zeta, g_mapping, kpt_thc.xi)
+        save_thc_factors(f"{checkpoint_basename}_isdf.h5", chi, zeta, g_mapping, kpt_thc.xi)
     momentum_map = build_momentum_transfer_mapping(kmf.cell, kmf.kpts)
     if cholesky is not None:
         cholesky_contiguous = make_contiguous_cholesky(cholesky)
@@ -907,10 +906,7 @@ def kpoint_thc_via_isdf(
         )
     start = time.time()
     if perform_bfgs_opt:
-        if save_checkpoints:
-            chkfile_name = f"{checkpoint_basename}_bfgs.h5"
-        else:
-            chkfile_name = None
+        bfgs_chkfile: str | None = f"{checkpoint_basename}_bfgs.h5" if save_checkpoints else None
         if use_batched_algos:
             opt_params, loss_bfgs = lbfgsb_opt_kpthc_l2reg_batched(
                 chi,
@@ -918,7 +914,7 @@ def kpoint_thc_via_isdf(
                 momentum_map,
                 g_mapping,
                 cholesky_contiguous,
-                chkfile_name=chkfile_name,
+                chkfile_name=bfgs_chkfile,
                 maxiter=bfgs_maxiter,
                 penalty_param=penalty_param,
                 batch_size=batch_size,
@@ -932,23 +928,22 @@ def kpoint_thc_via_isdf(
                 momentum_map,
                 g_mapping,
                 cholesky_contiguous,
-                chkfile_name=chkfile_name,
+                chkfile_name=bfgs_chkfile,
                 maxiter=bfgs_maxiter,
                 penalty_param=penalty_param,
                 disp_freq=(98 if verbose else -1),
             )
             info["loss_bfgs"] = loss_bfgs
         num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
-        chi, zeta = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+        chi, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+        zeta = np.array(zeta_list, dtype=object)
     if verbose:
         print("Time for BFGS {:.4f}".format(time.time() - start))
     start = time.time()
-    opt_params = ()
     if perform_adagrad_opt:
-        if save_checkpoints:
-            chkfile_name = f"{checkpoint_basename}_adagrad.h5"
-        else:
-            chkfile_name = None
+        adagrad_chkfile: str | None = (
+            f"{checkpoint_basename}_adagrad.h5" if save_checkpoints else None
+        )
         if use_batched_algos:
             opt_params, loss_ada = adagrad_opt_kpthc_batched(
                 chi,
@@ -956,13 +951,14 @@ def kpoint_thc_via_isdf(
                 momentum_map,
                 g_mapping,
                 cholesky_contiguous,
-                chkfile_name=chkfile_name,
+                chkfile_name=adagrad_chkfile,
                 maxiter=adagrad_maxiter,
                 batch_size=batch_size,
             )
             info["loss_adagrad"] = loss_ada
-        num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
-        chi, zeta = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+            num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
+            chi, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+            zeta = np.array(zeta_list, dtype=object)
     if verbose:
         print("Time for ADAGRAD {:.4f}".format(time.time() - start))
     result = KPointTHC(chi=chi, zeta=zeta, g_mapping=g_mapping, xi=kpt_thc.xi)

--- a/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
+++ b/src/openfermion/resource_estimates/pbc/thc/factorizations/thc_jax.py
@@ -114,8 +114,8 @@ def get_zeta_size(zeta: npt.NDArray) -> int:
 
 
 def unpack_thc_factors(
-    xcur: npt.NDArray, num_thc: int, num_orb: int, num_kpts: int, num_g_per_q: list[int]
-) -> tuple[npt.NDArray, list[npt.NDArray]]:
+    xcur: npt.NDArray | jax.Array, num_thc: int, num_orb: int, num_kpts: int, num_g_per_q: list[int]
+) -> tuple[npt.NDArray | jax.Array, list[npt.NDArray | jax.Array]]:
     """Unpack THC factors from flattened array used for reoptimization.
 
     Args:
@@ -314,9 +314,7 @@ def thc_objective_regularized_batched(
     """
     num_kpts = momentum_map.shape[0]
     num_g_per_q = [len(np.unique(gq)) for gq in gpq_map]
-    chi_packed, zeta_list = unpack_thc_factors(
-        np.asarray(xcur), num_thc, num_orb, num_kpts, num_g_per_q
-    )
+    chi_packed, zeta_list = unpack_thc_factors(xcur, num_thc, num_orb, num_kpts, num_g_per_q)
     chi_jax = jnp.asarray(chi_packed)
     # Normalization factor, no factor of sqrt as there are 4 chis in total when
     # building ERI.
@@ -522,7 +520,7 @@ def lbfgsb_opt_kpthc_l2reg(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
     return np.array(params), loss
 
 
@@ -664,7 +662,7 @@ def lbfgsb_opt_kpthc_l2reg_batched(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(gq)) for gq in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(params, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
     return np.array(params), loss
 
 
@@ -757,7 +755,7 @@ def adagrad_opt_kpthc_batched(
     if chkfile_name is not None:
         num_g_per_q = [len(np.unique(GQ)) for GQ in gpq_map]
         chi_out, zeta_out = unpack_thc_factors(x, num_thc, num_orb, num_kpts, num_g_per_q)
-        save_thc_factors(chkfile_name, chi_out, np.array(zeta_out, dtype=object), gpq_map)
+        save_thc_factors(chkfile_name, np.asarray(chi_out), np.array(zeta_out, dtype=object), gpq_map)
     return params, loss
 
 
@@ -935,7 +933,8 @@ def kpoint_thc_via_isdf(
             )
             info["loss_bfgs"] = loss_bfgs
         num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
-        chi, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+        chi_packed, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+        chi = np.asarray(chi_packed)
         zeta = np.array(zeta_list, dtype=object)
     if verbose:
         print("Time for BFGS {:.4f}".format(time.time() - start))
@@ -957,7 +956,10 @@ def kpoint_thc_via_isdf(
             )
             info["loss_adagrad"] = loss_ada
             num_g_per_q = [len(np.unique(GQ)) for GQ in g_mapping]
-            chi, zeta_list = unpack_thc_factors(opt_params, num_thc, num_mo, num_kpts, num_g_per_q)
+            chi_packed, zeta_list = unpack_thc_factors(
+                opt_params, num_thc, num_mo, num_kpts, num_g_per_q
+            )
+            chi = np.asarray(chi_packed)
             zeta = np.array(zeta_list, dtype=object)
     if verbose:
         print("Time for ADAGRAD {:.4f}".format(time.time() - start))

